src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsServerRouter/DiagnosticsServerRouter.cs

```diff
@@ -0,0 +1,841 @@
+using System.Diagnostics;
+using Microsoft.Extensions.Logging;
+using System.Net;
+
+namespace Microsoft.Diagnostics.NETCore.Client
+{
+    internal class RuntimeConnectTimeoutException : TimeoutException
+    {
+        public RuntimeConnectTimeoutException(int TimeoutMs)
+            : base(string.Format("No new runtime endpoints connected, waited {0} ms", TimeoutMs))
+        { }
+    }
+
+    internal class BackendStreamConnectTimeoutException : TimeoutException
+    {
+        public BackendStreamConnectTimeoutException(int TimeoutMs)
+            : base(string.Format("No new backend streams available, waited {0} ms", TimeoutMs))
+        { }
+    }
+
+    /// <summary>
+    /// Base class representing a Diagnostics Server router.
+    /// </summary>
+    internal class DiagnosticsServerRouter
+    {
+        protected readonly ILogger _logger;
+
+        public DiagnosticsServerRouter(ILogger logger)
+        {
+            _logger = logger;
+        }
+
+        public ILogger Logger
+        {
+            get { return _logger; }
+        }
+
+        public virtual void Start()
+        {
+            throw new NotImplementedException();
+        }
+
+        public virtual Task Stop()
+        {
+            throw new NotImplementedException();
+        }
+
+        public virtual void Reset()
+        {
+            throw new NotImplementedException();
+        }
+
+        public virtual Task<ConnectedRouter> ConnectRouterAsync(CancellationToken token)
+        {
+            throw new NotImplementedException();
+        }
+    }
+
+    /// <summary>
+    /// This class represent a TCP/IP server endpoint used when building up router instances.
+    /// </summary>
+    internal class TcpServerRouter : DiagnosticsServerRouter
+    {
+        protected string _tcpServerAddress;
+
+        protected ReversedDiagnosticsServer _tcpServer;
+        protected IpcEndpointInfo _tcpServerEndpointInfo;
+
+        public int RuntimeInstanceConnectTimeoutMs { get; set; } = 60000;
+        public int TcpServerConnectTimeoutMs { get; set; } = 5000;
+
+        public Guid RuntimeInstanceId
+        {
+            get { return _tcpServerEndpointInfo.RuntimeInstanceCookie; }
+        }
+
+        public int RuntimeProcessId
+        {
+            get { return _tcpServerEndpointInfo.ProcessId; }
+        }
+
+        public string TcpServerAddress
+        {
+            get { return _tcpServerAddress; }
+        }
+
+        protected TcpServerRouter(string tcpServer, int runtimeTimeoutMs, ILogger logger)
+            : base(logger)
+        {
+            _tcpServerAddress = tcpServer;
+            if (string.IsNullOrEmpty(_tcpServerAddress))
+            {
+                _tcpServerAddress = "127.0.0.1:0";
+            }
+            else
+            {
+                _tcpServerAddress = _tcpServerAddress.Replace("tcp://", "");
+                _tcpServerAddress = _tcpServerAddress.Replace(Uri.UriSchemeNetTcp, "");
```

Comment:
Does this need to do any other validation, e.g., determine whether the incoming address is malformed, check if the address is IPv4 vs IPv6, check is a user put `http://` rather than `tcp://`, etc.? Or is that handled upstream?

Reply:
The real parsing of this value will be taken care by IpcTcpSocketEndPoint::TryParseIPAddress that uses format host:port and that handles IPv4, IPv6 addresses as well as * (wildcard), binding to all interfaces. This code just get rid of any Uri schema prefixing that string (so we could differentiate, namedpipes, uds and tcp in the future). If the address will later be run through   IPAddress.TryParse and if that fails, Uri.TryCreate and if both fails, we won't be able to use the address and user will be informed and dsrouter will exit. We could probably drop the code above and leave the uri schema (if used), it will be dropped by TryParseIPAddress, just needs to make sure it will work with all scenarios.

Reply:
Create a function to normalize an endpoint string into format used internally (host:port), that function will take care of all parsing and error handling and is used from all location handling the endpoint address.

Reply:
Created a function to normalize an endpoint string into format used internally (host:port), that function will take care of all parsing and error handling and is used from all location handling the endpoint address.